File: src/simulation/space/tile/Tile.java
package simulation.space.tile;

import simulation.space.*;
import simulation.space.resource.*;

import java.util.*;
import java.util.function.Predicate;
import java.util.stream.Collectors;

import static simulation.Controller.session;

public class Tile {
    private int x, y;
    private MutableTileTagPool tagPool = new MutableTileTagPool();
    private Type type;
    private TectonicPlate plate;
    private MutableResourcePack resourcePack = new MutableResourcePack();
    /**
     * Resources which were added on this Tile during this turn. They are
     * stored here before the end of the turn.
     */
    private List<Resource> _delayedResources = new ArrayList<>();
    /**
     * Level of this Tile.
     */
    private int level;
    /**
     * Lowest level of this Tile which corresponds to the ground level.
     */
    private int secondLevel;
    private int temperature;
    private List<Tile> neighbours = null;
    private WindCenter windCenter = new WindCenter();

    public Tile(int x, int y) {
        this.x = x;
        this.y = y;
        updateTemperature();
        setType(Type.Normal, true);
    }

    public void setNeighbours(List<Tile> neighbours) {
        if (this.neighbours != null) {
            throw new SpaceError("Neighbours are already set");
        }
        this.neighbours = neighbours;
    }

    public ResourcePack getResourcePack() {
        return resourcePack;
    }

    public MutableTileTagPool getTagPool() {
        return tagPool;
    }

    public List<Resource> getResourcesWithMoved() {
        List<Resource> _l = new ArrayList<>(resourcePack.getResources());
        _l.addAll(_delayedResources);
        return _l;
    }

    /**
     * @return all Resources which are available from this Tile.
     */
    public List<Resource> getAccessibleResources() {
        List<Resource> accessibleResources = new ArrayList<>(resourcePack.getResources());
        getNeighbours().forEach(t -> accessibleResources.addAll(t.resourcePack.getResources()));
        return accessibleResources;
    }

    public List<Tile> getNeighbours(Predicate<Tile> predicate) {
        return neighbours.stream()
                .filter(predicate)
                .collect(Collectors.toList());
    }

    public List<Tile> getNeighbours() {
        return neighbours;
    }

    public Set<Tile> getTilesInRadius(int radius) {
        Set<Tile> tiles = new HashSet<>();
        Set<Tile> outer = new HashSet<>();
        if (radius > 0) {
            tiles.addAll(neighbours);
            outer.addAll(neighbours);
        }
        for (int i = 0; i < radius - 1; i++) {
            outer = outer.stream()
                    .flatMap(t -> t.neighbours.stream())
                    .filter(t -> !tiles.contains(t))
                    .collect(Collectors.toSet());
            tiles.addAll(outer);
        }
        return tiles;
    }

    public Set<Tile> getTilesInRadius(int radius, Predicate<Tile> predicate) {
        return getTilesInRadius(radius).stream()
                .filter(predicate)
                .collect(Collectors.toSet());
    }

    public TectonicPlate getPlate() {
        return plate;
    }

    public Type getType() {
        return type;
    }

    public int getLevel() {
        return level;
    }

    public int getSecondLevel() {
        return secondLevel;
    }

    public int getTemperature() {
        return temperature;
    }

    public Wind getWind() {
        return windCenter.getWind();
    }

    public int getX() {
        return x;
    }

    public int getY() {
        return y;
    }

    public void setPlate(TectonicPlate plate) {
        this.plate = plate;
    }

    public void setType(Type type, boolean updateLevel) {
        if (type == this.type) {
            return;
        }
        this.type = type;
        if (!updateLevel) {
            return;
        }
        switch (type) {
            case Mountain:
                level = 110;
                secondLevel = 110;
                break;
            case Normal:
                level = 100;
                secondLevel = 100;
                break;
            case Water:
                level = 85;
                secondLevel = 85;
                break;
        }
    }

    public void setLevel(int level) {
        this.level = level;
        this.secondLevel = level;
        type = Type.Normal;
        if (level >= 110) {
            type = Type.Mountain;
        }
        if (level < SpaceData.INSTANCE.getData().getDefaultWaterLevel()) {
            type = Type.Water;
        }
    }

    private void addResource(Resource resource) {
        if (resource.isEmpty()) {
            return;
        }
        resourcePack.add(resource);
    }

    /**
     * Adds resources which will be available on this Tile on the next turn.
     *
     * @param resource resource which will be added.
     */
    public void addDelayedResource(Resource resource) {
        if (resource.isEmpty()) {
            return;
        }
        _delayedResources.add(resource);
    }

    public void addDelayedResources(Collection<Resource> resources) {
        resources.forEach(this::addDelayedResource);
    }

    public void addDelayedResources(ResourcePack pack) {
        addDelayedResources(pack.getResources());
    }

    public void startUpdate() { //TODO wind blows on 2 neighbour tiles
        updateResources();
        windCenter.startUpdate();
        windCenter.useWind(resourcePack.getResources());
        updateTemperature();
        updateType();
    }

    public void middleUpdate() {
        _delayedResources.forEach(this::addResource);
        _delayedResources.clear();
        windCenter.middleUpdate(x, y);
    }

    public void finishUpdate() {
        windCenter.finishUpdate();
    }

    private void updateTemperature() {
        int start = SpaceData.INSTANCE.getData().getTemperatureBaseStart();
        int finish = SpaceData.INSTANCE.getData().getTemperatureBaseFinish();
        int size = SpaceData.INSTANCE.getData().getMapSizeX();
        int waterShift = 0;
        if (type == Type.Water) {
            waterShift -= 2;
            int deepness = Math.max(SpaceData.INSTANCE.getData().getDefaultWaterLevel() - level, 0);
            waterShift -= deepness;
        }
        int oldwaterShift = (type == Type.Water || type == Type.Ice ? -10 : 0);
        temperature = start + x * (finish - start) / size -
                Math.max(0, (level - 110) / 2) + waterShift;
    }

    private void updateResources() {
        List<Resource> deletedResources = new ArrayList<>();
        for (Resource resource: resourcePack.getResources()) {
            ResourceUpdateResult result = resource.update(this);
            if (!result.isAlive()) {
                deletedResources.add(resource);
            }
            addDelayedResources(result.getProduced());
        }
        resourcePack.removeAll(deletedResources);
    }

    private void updateType() {
        if (resourcePack.contains(session.world.getResourcePool().get("Water"))) {
            setType(Type.Water, false);
        }
        if (getType() == Type.Normal || getType() == Type.Woods || getType() == Type.Growth) {
            if (resourcePack.any(r -> r.getSimpleName().matches(".*Tree|Spruce"))) {
                setType(Type.Woods, false);
            } else if (resourcePack.any(r -> r.getGenome().getType() == Genome.Type.Plant)) {
                setType(Type.Growth, false);
            } else {
                setType(Type.Normal, false);
            }
        } else if (getType() == Type.Water) {
            addDelayedResource(session.world.getResourcePool().get("Vapour").copy(50 * session.resourceProportionCoefficient));
        }
        if (resourcePack.contains(session.world.getResourcePool().get("Water"))) {
            addDelayedResource(session.world.getResourcePool().get("Vapour").copy(50 * session.resourceProportionCoefficient));
        }
        if (type == Type.Water && temperature < -10) {
            type = Type.Ice;
            level = SpaceData.INSTANCE.getData().getDefaultWaterLevel();
        } else if (type == Type.Ice && temperature > 0) {
            type = Type.Water;
            level = secondLevel;
        }
    }

    public void levelUpdate() {//TODO works bad on Ice; wind should affect mountains mb they will stop grow
        for (int i = 0; i < (type == Type.Water ? 4 : (level > 105 && level < 120 ? 5 : 1)); i++) {
            distributeLevel();
        }
    }

    private void distributeLevel() {
        List<Tile> tiles = getNeighbours();
        tiles.sort(Comparator.comparingInt(tile -> tile.secondLevel));
        Tile lowest = tiles.get(0);
        if (level == 110 && lowest.level == 100) {
            int i = 0;
        }
        if (lowest.secondLevel + 1 < secondLevel) {
            setLevel(level - 1);
            lowest.setLevel(lowest.level + 1);
        }
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Tile tile = (Tile) o;
        return x == tile.x && y == tile.y;
    }

    @Override
    public String toString() {
        StringBuilder stringBuilder = new StringBuilder(String.format(
                "Tile %d %d, type=%s, temperature=%d, level=%d, resources:\n", x, y, getType(), temperature, level
                ));
        stringBuilder.append("Tags: ");
        for (TileTag tag : tagPool.getAll()) {
            stringBuilder.append(tag).append(" ");
        }
        stringBuilder.append('\n');
        for (Resource resource : resourcePack.getResources()) {
            stringBuilder.append(resource).append("\n");
        }
        return stringBuilder.toString();
    }

    public enum Type {
        Normal,
        Mountain,
        Water,
        Ice,
        Woods,
        Growth
    }
}

File: src/simulation/space/tile/Tile.kt
package simulation.space.tile

import simulation.Controller
import simulation.space.SpaceData.data
import simulation.space.SpaceError
import simulation.space.TectonicPlate
import simulation.space.resource.container.MutableResourcePack
import simulation.space.resource.Resource
import simulation.space.resource.container.ResourcePack
import java.util.*
import kotlin.math.max

class Tile(val x: Int, val y: Int, private val typeUpdater: TypeUpdater) {
    val tagPool = MutableTileTagPool()

    var type: Type? = null
        internal set

    var plate: TectonicPlate? = null

    private val _resourcePack = MutableResourcePack()

    val resourcePack: ResourcePack
        get() = _resourcePack

    /**
     * Resources which were added on this Tile during this turn. They are
     * stored here before the end of the turn.
     */
    private val _delayedResources: MutableList<Resource> = ArrayList()

    var level = 0
        internal set

    //Lowest level of this Tile which corresponds to the ground level.
    var secondLevel = 0
        private set

    var temperature = 0
        private set

    var neighbours = listOf<Tile>()
        set(value) {
            if (field.isNotEmpty())
                throw SpaceError("Neighbours are already set")
            field = value
        }

    init {
        updateTemperature()
        setType(Type.Normal, true)
    }

    private val windCenter = WindCenter()

    val resourcesWithMoved: List<Resource>
        get() {
            val resources: MutableList<Resource> = ArrayList(_resourcePack.resources)
            resources.addAll(_delayedResources)
            return resources
        }

    val accessibleResources: List<Resource>
        get() {
            val accessibleResources: MutableList<Resource> = ArrayList(_resourcePack.resources)
            accessibleResources.addAll(neighbours.flatMap { it._resourcePack.resources })
            return accessibleResources
        }

    fun getNeighbours(predicate: (Tile) -> Boolean) = neighbours.filter(predicate)

    fun getTilesInRadius(radius: Int): Set<Tile> {
        val tiles = mutableSetOf<Tile>()
        var outer = setOf<Tile>()

        if (radius > 0)
            tiles.addAll(neighbours)

        for (i in 0 until radius - 1) {
            outer = outer
                    .flatMap { it.neighbours }
                    .filter { !tiles.contains(it) }
                    .toSet()
            tiles.addAll(outer)
        }

        return tiles
    }

    fun getTilesInRadius(radius: Int, predicate: (Tile) -> Boolean) = getTilesInRadius(radius)
            .filter(predicate)
            .toSet()

    val wind: Wind
        get() = windCenter.wind

    fun setType(type: Type, updateLevel: Boolean) {
        if (type == this.type)
            return

        this.type = type

        if (!updateLevel)
            return

        when (type) {
            Type.Mountain -> {
                level = 110
                secondLevel = 110
            }
            Type.Normal -> {
                level = 100
                secondLevel = 100
            }
            Type.Water -> {
                level = data.seabedLevel
                secondLevel = data.seabedLevel
            }
            Type.Ice -> {
                level = data.defaultWaterLevel
            }
            else -> {}
        }
    }

    fun setLevel(level: Int) {
        this.level = level
        secondLevel = level
        type = Type.Normal

        if (level >= 110)
            type = Type.Mountain
        if (level < data.defaultWaterLevel)
            type = Type.Water
    }

    private fun addResource(resource: Resource) {
        if (resource.isEmpty)
            return
        _resourcePack.add(resource)
    }

    /**
     * Adds resources which will be available on this Tile on the next turn.
     *
     * @param resource resource which will be added.
     */
    fun addDelayedResource(resource: Resource) {
        if (resource.isEmpty)
            return
        _delayedResources.add(resource)
    }

    fun addDelayedResources(resources: Collection<Resource>) = resources.forEach { addDelayedResource(it) }

    fun addDelayedResources(pack: ResourcePack) = addDelayedResources(pack.resources)

    fun startUpdate() { //TODO wind blows on 2 neighbour tiles
        updateResources()
        windCenter.startUpdate()
        windCenter.useWind(_resourcePack.resources)
        updateTemperature()
        updateType()
    }

    fun middleUpdate() {
        _delayedResources.forEach { addResource(it) }
        _delayedResources.clear()
        windCenter.middleUpdate(x, y)
    }

    fun finishUpdate() {
        windCenter.finishUpdate()
    }

    private fun updateTemperature() {
        val start = data.temperatureBaseStart
        val finish = data.temperatureBaseFinish
        val size = data.mapSizeX
        var levelShift = 0
        if (type == Type.Water) {
            levelShift -= 2
            val deepness = max(data.defaultWaterLevel - level, 0)
            levelShift -= deepness
        } else {
            val elevation = max(level - data.defaultWaterLevel, 0)
            levelShift -= elevation / 3
        }
        temperature = start + x * (finish - start) / size + levelShift
    }

    private fun updateResources() {
        val deletedResources: MutableList<Resource> = ArrayList()
        for (resource in _resourcePack.resources) {
            val result = resource.update(this)
            if (!result.isAlive)
                deletedResources.add(resource)
            addDelayedResources(result.produced)
        }
        _resourcePack.removeAll(deletedResources)
    }

    private fun updateType() {
        typeUpdater.updateType(this)

        if (type == Type.Water)
            addDelayedResource(Controller.session.world.resourcePool.get("Vapour"))
    }

    fun levelUpdate() { //TODO works bad on Ice; wind should affect mountains mb they will stop growing
        for (i in 0 until if (type == Type.Water) 4 else if (level in 106..119) 5 else 1)
            distributeLevel()
    }

    private fun distributeLevel() {
        val tiles = neighbours.toMutableList()
        tiles.sortBy { it.secondLevel }
        val lowest = tiles[0]
        if (lowest.secondLevel + 1 < secondLevel) {
            setLevel(level - 1)
            lowest.setLevel(lowest.level + 1)
        }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val tile = other as Tile
        return x == tile.x && y == tile.y
    }

    override fun hashCode(): Int {
        var result = x
        result = 31 * result + y
        return result
    }

    override fun toString() = "Tile $x $y, type=$type, temperature=$temperature, level=$level, resources:\n" +
            "Tags: " + tagPool.all.joinToString("; ") +
            _resourcePack.resources.joinToString("\n", "\n", "\n")

    enum class Type {
        Normal, Mountain, Water, Ice, Woods, Growth
    }
}